### PR TITLE
Adding changes for support for making FlexVolume Working

### PR DESCRIPTION
**Pull request by errorsandwarnings**
This is related to issue in longhorn https://github.com/rancher/longhorn/issues/54 
Affects other storage drivers too like OpenEBS

Without this kubelet running as container in Rancher 1.6.5 can not access the directory /usr/libexec/kubernetes/kubelet-plugins and hence the storage volumes like longhorn do not work properly.

### CODE REVIEW

**Comment by yasker** [comment]
The default volume plugin directory `/usr/libexec/kubernetes/kubelet-plugins` doesn't work on certain distro e.g. CoreOS. It's better if we change it to `/var/lib/kubelet/volumeplugins` (which is already bind-mounted) and add `--volume-plugin-dir=/var/lib/kubelet/volumeplugins` to the kubelet command line instead.

Though then in Kubernetes 1.9+, an explicit configuration for `FLEXVOLUME_DIR` would be needed for Longhorn (and other flexvolume plugins) installation.

**Comment by errorsandwarnings** [comment]
@yasker : Will the above for now work for atleast OS other then CoreOS ?

**Comment by yasker** [comment]
I think it breaks atomic OS as well.

There is not really a downside to use a non-default location except user will have to manually configure flexvolume dir for 1.9. Storage vendors has documents to explains it. Excluding some OS is a bigger issue.

Also, Longhorn is moving to CSI for Kubernetes 1.10, so it will only be the issue with 1.9.

**Comment by errorsandwarnings** [comment]
@yasker 
Yeah, but seems like this is also an issue for other things storage volumes like OpenEBS. We can not expect everyone to move to Kubernetes 1.10, backwards compatability should be taken care of as well.

**Comment by yasker** [comment]
I don't understand your points.

As a container orchestrator, we cannot break certain OS's ability to run Kubernetes with Flexvolume.

What issue does it bring to OpenEBS? If OpenEBS doesn't support customized plugin location, then they need to fix it.

And I don't see the backward compatibility issue here as well. It wasn't supported before. The user will need to supply the same location after v1.9 for the flexvolume plugin to install properly. So the compatibility is fine.

**Comment by errorsandwarnings** [comment]
@yasker : I agree with the point that we can not break certain OS's functionality to run kubernetes. But what is the solution then, keep changing yml files manually ?

**Comment by errorsandwarnings** [comment]
@yasker : All I want is a solution to be on table for people running rancher, installing kubernetes 1.9.5 and working with longhorn.

File: infra-templates/k8s/45/docker-compose.yml.tpl
Comment (yasker) [inline infra-templates/k8s/45/docker-compose.yml.tpl]: Instead of adding this bind-mount here, add e.g. `--volume-plugin-dir=/var/lib/kubelet/volumeplugins` to the command line of `kubelet` above, e.g. line 34. Kubelet will use `/var/lib/kubelet/volumeplugins` instead and that directory has already been bind-mounted.

**Review by yasker — CHANGES_REQUESTED** [review]
We cannot break CoreOS or AtomicOS, so let's use `kubelet --volume-plugin-dir` instead of bind mounting default directory.

**Comment by errorsandwarnings** [comment]
@yasker : will the same need to be done for kubelet-unschedulable ?

**Review by errorsandwarnings — COMMENTED** [review]
Commit Added

**Review by errorsandwarnings — COMMENTED** [review]
Does this look okay now ?

**Comment by yasker** [comment]
@errorsandwarnings looks good now. You need to squash the change into one commit.

Have you tested it?

**Comment by kingsd041** [comment]
@yasker 
Tested with rancher-server version-v1.6.15 with kubernetes v1.9.5
I customized a catalog and added the `--volume-plugin-dir=/var/lib/kubelet/volumeplugins` parameter to docker-compose.yml.tpl
https://github.com/kingsd041/catalog-0412/blob/master/infra-templates/k8s/46/docker-compose.yml.tpl#L36
Unfortunately, the pod still can't start. 
```
Events:
  Type     Reason                 Age   From                                                 Message
  ----     ------                 ----  ----                                                 -------
  Normal   Scheduled              2m    default-scheduler                                    Successfully assigned volume-test to ip-172-31-31-67.eu-west-1.compute.internal
  Normal   SuccessfulMountVolume  2m    kubelet, ip-172-31-31-67.eu-west-1.compute.internal  MountVolume.SetUp succeeded for volume "default-token-lrw2g"
  Warning  FailedMount            17s   kubelet, ip-172-31-31-67.eu-west-1.compute.internal  Unable to mount volumes for pod "volume-test_default(ebc53a84-3e54-11e8-88b7-028894561beb)": timeout expired waiting for volumes to attach/mount for pod "default"/"volume-test". list of unattached/unmounted volumes=[volv]
```
```
root@ip-172-31-20-161:/# ps aux | grep kubelet
root     17888  2.6  2.5 610908 103684 ?       Ssl  13:14   1:15 kubelet --kubeconfig=/etc/kubernetes/ssl/kubeconfig --allow-privileged=true --register-node=true --cloud-provider=rancher --healthz-bind-address=0.0.0.0 --cluster-dns=10.43.0.10 --fail-swap-on=false --cluster-domain=cluster.local --network-plugin=cni --cni-conf-dir=/etc/cni/managed.d --anonymous-auth=false --volume-plugin-dir=/var/lib/kubelet/volumeplugins --client-ca-file=/etc/kubernetes/ssl/ca.pem --pod-infra-container-image=rancher/pause-amd64:3.0 --cgroup-driver=cgroupfs --hostname-override ip-172-31-20-161.eu-west-1.compute.internal
```

longhorn-flexvolume-driver
```
> kubectl logs longhorn-flexvolume-driver-w4jrv -n longhor-system
Error from server (NotFound): namespaces "longhor-system" not found
> kubectl logs longhorn-flexvolume-driver-w4jrv -n longhorn-system
+ echo Dependency checking
++ nsenter --mount=/host/proc/1/ns/mnt -- nsenter -t 1 -n findmnt --version
Dependency checking
+ OUT='findmnt from util-linux 2.27.1'
++ nsenter --mount=/host/proc/1/ns/mnt -- nsenter -t 1 -n curl --version
+ OUT='curl 7.47.0 (x86_64-pc-linux-gnu) libcurl/7.47.0 GnuTLS/3.4.10 zlib/1.2.8 libidn/1.32 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP UnixSockets '
++ nsenter --mount=/host/proc/1/ns/mnt -- nsenter -t 1 -n blkid -v
Detecting backend service IP for longhorn-backend
+ OUT='blkid from util-linux 2.27.1  (libblkid 2.27.0, 02-Nov-2015)'
+ exit 0
Backend service IP for longhorn-backend is 10.43.68.99
Flexvolume driver installed
```

longhorn-flexvolume-driver-deployer
```
kubectl logs longhorn-flexvolume-driver-deployer-85dd94b9bc-qqt7x -n longhorn-system
INFO[0000] cannot find volumePluginDir key in node config, assume it's default
INFO[0000] Install Flexvolume to Kubernetes nodes directory /usr/libexec/kubernetes/kubelet-plugins/volume/exec/
```

longhorn-manager
```
kubectl log longhorn-manager-j7d2h -n longhorn-system
W0412 22:00:45.865927   17048 cmd.go:392] log is DEPRECATED and will be removed in a future version. Use logs instead.
ERROR: logging before flag.Parse: I0412 13:20:11.818251       1 controller_utils.go:1041] Waiting for caches to sync for longhorn datastore controller
ERROR: logging before flag.Parse: I0412 13:20:11.920165       1 controller_utils.go:1048] Caches are synced for longhorn datastore controller
INFO[0000] Start Longhorn replica controller
ERROR: logging before flag.Parse: I0412 13:20:11.920346       1 controller_utils.go:1041] Waiting for caches to sync for longhorn replicas controller
INFO[0000] Start Longhorn engine controller
ERROR: logging before flag.Parse: I0412 13:20:11.920377       1 controller_utils.go:1041] Waiting for caches to sync for longhorn engines controller
INFO[0000] Start Longhorn volume controller
ERROR: logging before flag.Parse: I0412 13:20:11.920410       1 controller_utils.go:1041] Waiting for caches to sync for longhorn engines controller
ERROR: logging before flag.Parse: I0412 13:20:11.940709       1 controller.go:407] Starting provisioner controller 3aca47ec-3e54-11e8-b790-02889403f502!
INFO[0000] Listening on 10.42.248.15:9500
ERROR: logging before flag.Parse: I0412 13:20:12.020532       1 controller_utils.go:1048] Caches are synced for longhorn engines controller
ERROR: logging before flag.Parse: I0412 13:20:12.020556       1 controller_utils.go:1048] Caches are synced for longhorn engines controller
ERROR: logging before flag.Parse: I0412 13:20:12.020533       1 controller_utils.go:1048] Caches are synced for longhorn replicas controller
ERROR: logging before flag.Parse: I0412 13:24:35.546569       1 controller.go:1084] scheduleOperation[lock-provision-default/longhorn-volv-pvc[d7e764d0-3e54-11e8-88b7-028894561beb]]
ERROR: logging before flag.Parse: I0412 13:24:35.583440       1 leaderelection.go:156] attempting to acquire leader lease...
ERROR: logging before flag.Parse: E0412 13:24:35.609187       1 leaderelection.go:273] Failed to update lock: Operation cannot be fulfilled on persistentvolumeclaims "longhorn-volv-pvc": the object has been modified; please apply your changes to the latest version and try again
```

Others:
If I add the `/usr/libexec/kubernetes/kubelet-plugins:/usr/libexec/kubernetes/kubelet-plugins` parameter to the catalog, the pod can start successfully.
https://github.com/kingsd041/k8s2-catalog/blob/master/infra-templates/k8s/46/docker-compose.yml.tpl#L70

**Comment by niusmallnan** [comment]
@kingsd041 Can you try a customize Flexvolume directory, like:
```
- name: FLEXVOLUME_DIR
  value: "/var/lib/kubelet/volumeplugins"
```

**Comment by kingsd041** [comment]
@niusmallnan  thanks
@yasker Sorry, just now I ignored FLEXVOLUME_DIR. The pod can boot successfully after I modify the FLEXVOLUME_DIR directory.
Test passed!

**Comment by yasker** [comment]
Thanks @kingsd041 @niusmallnan !

**Review by yasker — APPROVED** [review]

**Comment by yasker** [comment]
@errorsandwarnings I still need the commits squashed into one (instead of 7), after that we should able to merge it.

**Review by yasker — CHANGES_REQUESTED** [review]
Need to squash into one commit.

**Comment by deniseschannon** [comment]
@errorsandwarnings any chance you can squash this into 1 commit asap? we're planning on shipping a release with this k8s version in it and if this doesn't make the cut off, it'll need to wait until our next release cycle.

**Pull request merged**
